src/rvsdg/optimize_direct_jumps.rs

```diff
@@ -99,51 +119,105 @@ impl SimpleCfgFunction {
         }
     }
 
+    fn fixup_phis(block: &mut BasicBlock, label_mapping: &HashMap<String, String>) {
+        for code in &mut block.instrs {
+            if let Instruction::Value {
+                op: ValueOps::Phi,
+                ref mut labels,
+                ..
+            } = code
+            {
+                labels.iter_mut().for_each(|node| {
+                    if let Some(label) = label_mapping.get(node) {
+                        *node = label.to_string();
+                    }
+                });
+            }
+        }
+    }
+
     // this function looks for all CFG fragments of the form
     // source node --(source edge)-> empty node --(target edge)-> target node
     // (where there is only one target edge and the empty node has no instructions)
     // and changes the source edge to point to the target node
     fn collapse_empty_blocks(mut self) -> SimpleCfgFunction {
-        let target_edges: Vec<EdgeIndex> = self
-            .graph
-            .node_indices()
-            .filter(|node| {
-                self.graph[*node].instrs.is_empty() && self.graph[*node].footer.is_empty()
-            })
-            .filter_map(|node| {
-                let outgoing: Vec<_> = self
-                    .graph
-                    .edges_directed(node, Direction::Outgoing)
-                    .collect();
-
-                match outgoing.as_slice() {
-                    [edge] => Some(edge.id()),
-                    _ => None,
-                }
-            })
-            .collect();
-
-        for target_edge in target_edges {
-            let (empty, target) = self.graph.edge_endpoints(target_edge).unwrap();
-
-            let source_edges: Vec<EdgeIndex> = self
+        // loop over every non-empty block, and look at its parents
+        for block in self.graph.node_indices().collect::<Vec<_>>() {
+            // find parents
+            let parents: Vec<NodeIndex> = self
                 .graph
-                .edges_directed(empty, Direction::Incoming)
-                .map(|edge| edge.id())
+                .edges_directed(block, Direction::Incoming)
+                .map(|edge| edge.source())
                 .collect();
+            let new_parents = parents
+                .iter()
+                .map(|parent| self.get_parent_skipping_empty(*parent))
+                .collect::<HashSet<_>>();
+            let parent_mapping = parents
+                .iter()
+                .map(|parent| {
+                    (
+                        self.graph[*parent].name.to_string(),
+                        self.graph[self.get_parent_skipping_empty(*parent)]
+                            .name
+                            .to_string(),
+                    )
+                })
+                .collect::<HashMap<_, _>>();
+            // if we have fewer parents, phi nodes may get messed up, so avoid the optimization
+            if new_parents.len() == parents.len() {
```

Comment:
This makes the optimization markedly less effective. If both parents become the same, that means we have a diamond pattern where the sides are empty, and we 100% want to simplify that. You should make `fixup_phis` handle the case where phi instruction become id instructions.

Reply:
Actually, if both parents are empty your `select` optimization should fire. So this case should be handled by egglog.